src/tamm/tensor_impl.hpp

```diff
@@ -1153,8 +1157,16 @@ class DenseTensorImpl: public TensorImpl<T> {
             upcxx::global_ptr<uint8_t> remote_addr =
               gptrs_[t.rank] + (t.offset * elem_sz) + (tile_offset * elem_sz);
             uint8_t* local_addr = ((uint8_t*) buf) + (next++ * elem_sz);
-            upcxx::rput(local_addr, remote_addr, elem_sz, upcxx::operation_cx::as_promise(p));
+
+            all_puts[t.rank].first.push_back(local_addr);
+            all_puts[t.rank].second.push_back(remote_addr);
```

Comment:
@jcorbino Thanks for making this improvement to enable aggregation, it should make an enormous difference to the OFF-node RMA performance.

With that in place, I'd suggest one additional refinement, to reduce overhead for the ON-node case:

```suggestion
            if (remote_addr.is_local()) { // on-node : utilize lightweight shared-memory bypass
                *((T*)remote_addr.local()) = *(T*)local_addr;
            } else { // off-node : aggregated RMA using rput_regular below
                all_puts[t.rank].first.push_back(local_addr);
                all_puts[t.rank].second.push_back(remote_addr);
            }
```

This won't change the network behavior at all, but it should reduce the overhead of `std::vector` manipulation for on-node targets that don't benefit from aggregation (because they are satisfied through shared-memory bypass directly to the memory system, instead of network packets). In particular, with this change single-node runs will automatically skip the vector manipulation path entirely and use very lightweight load/store instructions.

Of course a symmetric refinement can also be made to the `get_raw` function.

Reply:
Thanks @bonachea , it certainly looks way better now! I completely forgot to optimize for shared-memory bypass.